src/librustc/hir/map/mod.rs

```diff
@@ -455,11 +455,20 @@ impl<'hir> Map<'hir> {
             Node::AnonConst(_) => {
                 BodyOwnerKind::Const
             }
+            Node::Variant(&Spanned { node: VariantKind { data: VariantData::Tuple(..), .. }, .. }) |
+            Node::StructCtor(..) |
+            Node::Item(&Item { node: ItemKind::Fn(..), .. }) |
+            Node::TraitItem(&TraitItem { node: TraitItemKind::Method(..), .. }) |
+            Node::ImplItem(&ImplItem { node: ImplItemKind::Method(..), .. }) => {
+                BodyOwnerKind::Fn
+            }
             Node::Item(&Item { node: ItemKind::Static(_, m, _), .. }) => {
                 BodyOwnerKind::Static(m)
             }
-            // Default to function if it's not a constant or static.
-            _ => BodyOwnerKind::Fn
+            Node::Expr(&Expr { node: ExprKind::Closure(..), .. }) => {
+                BodyOwnerKind::Closure
+            }
+            node => bug!("{:#?} is not a body node", node),
```

Comment:
This currently breaks Clippy in [`clippy_lints/src/utils/mod.rs`](https://github.com/rust-lang/rust-clippy/blob/9d5b14864827d8b4655842b29a1bd803e15e3f17/clippy_lints/src/utils/mod.rs#L65-L71):
```rust
pub fn in_constant(cx: &LateContext<'_, '_>, id: NodeId) -> bool {
    let parent_id = cx.tcx.hir().get_parent(id);
    match cx.tcx.hir().body_owner_kind(parent_id) {
        hir::BodyOwnerKind::Fn | hir::BodyOwnerKind::Closure => false,
        hir::BodyOwnerKind::Const | hir::BodyOwnerKind::Static(..) => true,
    }
}
```
This is because it is called on an enum variant here: [`clippy_lints/src/non_copy_const.rs`](https://github.com/rust-lang/rust-clippy/blob/9d5b14864827d8b4655842b29a1bd803e15e3f17/clippy_lints/src/non_copy_const.rs#L196-L198). `self.get(id)` returns `Node::Item(Item { node: ItemKind::Enum(..), .. })`

Should this function really produce an ICE if the `node` isn't one of the above. I think the previous approach of just assuming it is a function is also wrong. Maybe another variant for `BodyOwnerKind` would be better (`BodyOwnerKind::Other`?)

I would be happy to open a PR if this refactor should be done. Otherwise I'll try to fix this in Clippy.

cc @oli-obk

Reply:
We could make it return an `Option`, but I'm not sure if that is desirable. We didn't want to know whether something was  a function, closure or something else, we just wanted to know whether it is a constant or static.

In the end I decided to just check for const/static manually: https://github.com/rust-lang/rust-clippy/pull/3685/files